grp/basic.gd

```diff
@@ -343,64 +343,125 @@ end );
-##  is a multiple of 4.
+##  <Ref Func="DicyclicGroup"/> constructs the dicyclic group of size <A>n</A>
+##  in the category given by the #  filter <A>filt</A>. Here, <A>n</A> is a
+##  multiple of 4. <Ref Func="QuaternionGroup"/> is a synonym for
+##  <Ref Func="DicyclicGroup" /> provided for backward compatibility, but will
+##  print a warning if <A>n</A> is not a power of <M>2</M>.
 ##  If <A>filt</A> is not given it defaults to <Ref Func="IsPcGroup"/>.
 ##  For more information on possible values of <A>filt</A> see section
 ##  (<Ref Sect="Basic Groups"/>).
 ##  Methods are also available for permutation and matrix groups (of minimal
 ##  degree and minimal dimension in coprime characteristic).
 ##  <P/>
 ##  <Example><![CDATA[
-##  gap> QuaternionGroup(32);
-##  <pc group of size 32 with 5 generators>
+##  gap> DicyclicGroup(24);
+##  <pc group of size 24 with 4 generators>
 ##  gap> g:=QuaternionGroup(IsMatrixGroup,CF(16),32);
 ##  Group([ [ [ 0, 1 ], [ -1, 0 ] ], [ [ E(16), 0 ], [ 0, -E(16)^7 ] ] ])
 ##  ]]></Example>
 ##  </Description>
 ##  </ManSection>
 ##  <#/GAPDoc>
 ##
-BindGlobal( "QuaternionGroup", function ( arg )
+BindGlobal( "GRPLIB_DicyclicParameterCheck",
+function(args, quaternion)
+    local res, size;
+
+    res := [];
+
+    if Length(args) = 1 then
+        res[1] := IsPcGroup;
+        res[2] := args[1];
+        size := res[2];
+    elif Length(args) = 2 then
+        res[1] := args[1];
+        res[2] := args[2];
+        size := res[2];
+    elif Length(args) = 3 then
+        res[1] := args[1];
+        res[2] := args[2];
+        res[3] := args[3];
+        size := res[3];
+
+        if not IsField(res[2]) then
+            ErrorNoReturn("usage: <field> must be a field");
+        fi;
+    else
+        ErrorNoReturn("usage: DicyclicGroup( [<filter>, [<field>, ] ] <size> )");
+    fi;
 
-  if Length(arg) = 1  then
-    return QuaternionGroupCons( IsPcGroup, arg[1] );
-  elif IsOperation(arg[1]) then
-    if Length(arg) = 2  then
-      return QuaternionGroupCons( arg[1], arg[2] );
-    elif Length(arg) = 3  then
-      # some filters require extra arguments, e.g. IsMatrixGroup + field
-      return QuaternionGroupCons( arg[1], arg[2], arg[3] );
+    if not IsFilter(res[1]) then
+        ErrorNoReturn("usage: <filter> must be a filter");
     fi;
-  fi;
-  Error( "usage: QuaternionGroup( [<filter>, ]<size> )" );
 
-end );
+    if not (IsPosInt(size) and (size mod 4 = 0)) then
+        ErrorNoReturn("usage: <size> must be a positive integer divisible by 4");
+    fi;
+
+    if not IsPrimePowerInt(size) or size < 8 then
+        if quaternion = "error" then
+            ErrorNoReturn("usage: <size> must be a power of 2 and at least 8");
+        elif quaternion = "warn" then
+            Info(InfoWarning, 1, "Warning: QuaternionGroup called with <size> ", size,
+                                 " which is less than 8, or not a power of 2.");
```

Comment:
Optional: we could be less mysterious and just tell the user explicitly what the problem is (also, I'd remove the trailing period, which is unusual in info messages):
```gap
elif quaternion = "warn" then
  if not IsPrimePowerInt(size) then
     Info(InfoWarning, 1, "Warning: QuaternionGroup called with <size> ", size, " which is not a power of 2");
  else
     Info(InfoWarning, 1, "Warning: QuaternionGroup called with <size> ", size, " which is less than 8");
  fi;
```